### PR TITLE
Improve pickler efficiency

### PR DESCRIPTION
**Pull request by jjlilley**
Summary:
When writing common types in the pickler, it's common to perform
the write as multiple calls to the backend write_ function.

For instance, writing a larger int64 incurs 3 calls. Each double
incurs 9 calls. Smaller ints incur 2 calls.

The issue is that the write_ call often incurs per-call bounds
check overhead. And if we're only writing a couple of bytes, it's
silly to do so one byte at a time.

This change simply introduces a 2-argument and 3-argument push<>()
call, so that if there are multiple small fixed calls in a row, we
can change:
  push<PickleOpCode>(PickleOpCode::BININT1);
  push<uint8_t>(number);
Into:
  push<PickleOpCode, uint8_t>(PickleOpCode::BININT1, number);

The result is still type-safe, but it results in fewer backend
write calls and bounds checks.

Additionally, we avoid re-evaluating the ivalue type in a typed
list where the value is already known.

For a benchmark with a 1M int tensor, the net effect is a performance
win of 25% over the total serialization time.

Test Plan: buck test mode/dev-nosan caffe2/test/...

Differential Revision: D17774833

### CODE REVIEW

**Comment by pritamdamania87** [comment]
> For a benchmark with a 1M int tensor, the net effect is a performance
> win of 25% over the total serialization time.

Could you add the code used for benchmarking in this PR? Also, just to be sure this perf gain was in `opt` mode correct?

**Comment by jjlilley** [comment]
Thanks for looking.
 - The benchmark that I used is checked into fb experimental/jeremyl/c2/SerializationBench.cpp
   I can add them to the pytorch tree, but it wasn't entirely straightforward where to do so... do you have a suggestion for location? (also, it seems like we're not supposed to rely on folly in pytorch proper?)
 - Yes, it's @mode/opt :)  I have pprof traces that I can send out of band (put them on a chat group)

File: torch/csrc/jit/pickler.h [hunk 1]
```diff
@@ -193,6 +194,31 @@ class Pickler {
     writer_(begin, sizeof(T));
   }
 
+  // Specialize 2 & 3 argument version of push as well, intended for small
+  // values. The motivation is that writer_ calls often incur per-call
+  // bounds/size checks. If we're just writing a few bytes, combine the sizes on
+  // the stack before sending off to the writer.
+  template <typename A, typename B>
```
Comment (resistor) [inline torch/csrc/jit/pickler.h]: Would it be possible to instead define a version that takes an arbitrary number of arguments using a variadic template?

Comment (jjlilley) [reply torch/csrc/jit/pickler.h]: Let me think about variadic templates idea - trying to think offhand how to handle the common buffer between the args cleanly. It's probably possible, but my question is whether the resulting code is going to be more readable (or more obfuscated) with the variadic templates idiom applied.

Comment (zdevito) [reply torch/csrc/jit/pickler.h]: I feel like a standard buffering approach to calling `write_` where we used a pickler-local 1kb buffer would get us the same or potentially more wins without requiring somewhat invasive API changes.

Comment (jjlilley) [reply torch/csrc/jit/pickler.h]: Thanks for the feedback.

I don't think double-buffering necessarily makes sense here. The actual benchmark (and prod code) is just appending to a std::vector<char> behind the scenes. The main costs of doing this are effectively
  a) the branching related to bounds-checking and variable length memcpy.
  b) The std::function<> call overhead, effectively equivalent to a virtual function call

Double-buffering addresses (b) but not really (a).
The uploaded patch effectively addresses both issues, and it doesn't seem to be particularly complex.

fwiw, here's the resulting disassembly of the new pushInt, where we converted the 3 push<> calls to 1 for int64:
    P116779999

**Comment by jjlilley** [comment]
fwiw, I hacked up a double-buffering alternative here. (fb link is D17847174, it's benchmarkable by not reviewable yet)

For my 1 million int64 benchmark, performance is identical compared with this push<X,Y> approach, and both approaches save ~42% vs what's checked in. The double-buffering approach is admittedly slightly simpler to read.

@resistor , @zdevito  - if you prefer the double-buffering approach, I can clean up that version and send out for a real review. Either way, I just want to squeeze some cycles out of these codepaths. :)

**Comment by resistor** [comment]
@jjlilley I prefer the double-buffered version - it seems more future proof as the pickler is modified & extended.

**Comment by jjlilley** [comment]
ok, I'm abandoning this change in favor of https://github.com/pytorch/pytorch/pull/27720 which does the double-buffering.

**Pull request closed without merging**